src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java

```diff
@@ -81,26 +81,29 @@ public static boolean checkLdapFilter(Class<?> clz) {
      * The given factory class is rejected if any of these two filters reject
      * it, or if none of them allow it.
      *
-     * @param clz objects factory class
+     * @param serialClass objects factory class
      * @return true - if the factory is allowed to be instantiated; false - otherwise
      */
-    public static boolean checkRmiFilter(Class<?> clz) {
-        return checkInput("rmi", () -> clz);
+    public static boolean checkRmiFilter(Class<?> serialClass) {
+        return checkInput(RMI_FILTER, () -> serialClass);
     }
 
-    private static boolean checkInput(String scheme, FactoryInfo factoryInfo) {
-        Status result = switch(scheme) {
-            case "ldap" -> LDAP_FILTER.checkInput(factoryInfo);
-            case "rmi" -> RMI_FILTER.checkInput(factoryInfo);
-            default -> Status.UNDECIDED;
-        };
-        return switch(result) {
+    private static boolean checkInput(ConfiguredFilter filter, FactoryInfo serialClass) {
+        var globalFilter = GLOBAL_FILTER.filter();
+        var specificFilter = filter.filter();
+
+        Status globalResult = globalFilter.checkInput(serialClass);
+        if (filter == GLOBAL_FILTER) {
+            return globalResult == Status.ALLOWED;
+        }
+
```

Comment:
If I'm not mistaken there's  no point in checking the specific filter if the global filter state is REJECTED. So instead of switching on the specificResult below, maybe you should change the logic to switch on the globalResult instead and only call the specific filter if needed?

Reply:
> If I'm not mistaken there's no point in checking the specific filter if the global filter state is REJECTED. So instead of switching on the specificResult below, maybe you should change the logic to switch on the globalResult instead and only call the specific filter if needed?

Yes - there is no point, and that will reduce number of `checkInput` calls on a specific filter, if it is not the global one. That's how it will look like:
```
    private static boolean checkInput(ConfiguredFilter filter, FactoryInfo serialClass) {
        var globalFilter = GLOBAL_FILTER.filter();
        var specificFilter = filter.filter();
        Status globalResult = globalFilter.checkInput(serialClass);

        // Check if a specific filter is the global one
        if (filter == GLOBAL_FILTER) {
            return globalResult == Status.ALLOWED;
        }
        return switch (globalResult) {
            case ALLOWED -> specificFilter.checkInput(serialClass) != Status.REJECTED;
            case REJECTED -> false;
            case UNDECIDED -> specificFilter.checkInput(serialClass) == Status.ALLOWED;
        };
    }
```

The `if (filter == GLOBAL_FILTER) {` check can be also removed but without it the `checkInput` will be called twice on global filter for the case where `specific == global`, ie call from the `checkGlobalFilter`.

Reply:
The code above LGTM. An alternative could be:

```
    private static boolean checkInput(ConfiguredFilter filter, FactoryInfo serialClass) {
        var globalFilter = GLOBAL_FILTER.filter();
        var specificFilter = filter.filter();
        Status globalResult = globalFilter.checkInput(serialClass);

        return switch (globalResult) {
            case ALLOWED -> filter == GLOBAL || specificFilter.checkInput(serialClass) != Status.REJECTED;
            case REJECTED -> false;
            case UNDECIDED -> specificFilter.checkInput(serialClass) == Status.ALLOWED;
        };
    }
```

but I believe your proposed code reads better.

Reply:
Thanks - pushed as 528489b